### PR TITLE
New LAN Automation Module

### PR DESCRIPTION
**Pull request by maaliaj**
Added LAN Automation module with the following features:
  - Configuring LAN Automation sessions in Cisco Catalyst Center for automated
    discovery of devices and their integration into the network.
  - Updating LAN Automation device configurations, including loopback addresses,
    hostnames, and link configurations.
  - Automatically stopping an ongoing LAN Automation session based on conditions
    like timeout or discovery device list completion, without explicitly calling
    the stop API.
  - Stopping a LAN Automation session.
  - PnP device authorization.

## Type of Change
- [ ] Bug fix
- [X] New feature
- [ ] Breaking change
- [ ] Documentation update

## Checklist
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [X] Tasks are idempotent (can be run multiple times without changing state)
- [X] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [X] Playbooks are modular and reusable
- [X] Handlers are used for actions that need to run on change

## Documentation
- [X] All options and parameters are documented clearly.
- [X] Examples are provided and tested.
- [X] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/lan_automation_workflow_manager.py [hunk 1]
```diff
@@ -0,0 +1,2677 @@
...
+                else:
+                    pending_authorization = False
+                    self.msg = ("PnP authorization is set to True, but no devices were given for authorization or "
+                                "discovery. Please authorize devices manually in the PnP page in Catalyst Center.")
+                    self.set_operation_result("failed", False, self.msg, "ERROR")
+
+            self.log("Waiting for 30 seconds...")
+            time.sleep(self.params.get("dnac_task_poll_interval", 30))
+
+        if self.status != "success":
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2078]: elapsed_time = time.time() - start_time

        "LAN automation did not complete within the expected time of {:.2f} seconds. Consider stopping the "

File: plugins/modules/lan_automation_workflow_manager.py [hunk 2]
```diff
@@ -0,0 +1,2677 @@
...
+        payload_map = {
+            "loopback_update": "loopbackUpdateDeviceList",
+            "hostname_update": "hostnameUpdateDevices",
+            "link_add": "linkUpdate",
+            "link_delete": "linkUpdate"
+        }
+
+        feature = feature_map.get(update_type)
+        payload_key = payload_map.get(update_type)
+
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2119]: self.log("Preparing to call LAN automation API for update type '{}': {}".format(update_type, payload), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 3]
```diff
@@ -0,0 +1,2677 @@
...
+            empty dictionary. The logs can be useful for debugging and monitoring the state of
+            the automation tasks.
+        """
+        response = None
+        try:
+            response = self.dnac._exec(
+                family="lan_automation",
+                function="lan_automation_log",
+                op_modifies=False
+            )
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2227]: self.log("Received API response from 'lan_automation_log': {}".format(response), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 4]
```diff
@@ -0,0 +1,2677 @@
...
+                        device_ids.append(device_id)
+                    else:
+                        self.log(
+                            "Device {} is NOT in Pending Authorization state; current state: {}".format(serial_number,
+                                                                                                        state), "INFO")
+                else:
+                    self.log("Serial number {} is not in the provided device_serials list".format(serial_number),
+                             "INFO")
+
+        self.log("This is serial to id mapping for devices in 'Pending Authorization': {}".format(serial_to_id_map),
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2287]: self.log("Serial to ID mapping for devices in 'Pending Authorization': {}".format(serial_to_id_map), "DEBUG")
    self.log("Device IDs list for devices in 'Pending Authorization': {}".format(device_ids), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 5]
```diff
@@ -0,0 +1,2677 @@
...
+
+        self.log("Attempting to authorize devices with IDs: {}".format(device_ids), "DEBUG")
+        authorized_devices = self.authorize_pnp_devices(device_ids, serial_to_id_map)
+
+        if not authorized_devices:
+            self.log("Authorization failed for one or more devices", "ERROR")
+            return []
+
+        authorized_serials = [serial for serial, id in serial_to_id_map.items() if id in authorized_devices]
+
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2313]: new line not needed

File: plugins/modules/lan_automation_workflow_manager.py [hunk 6]
```diff
@@ -0,0 +1,2677 @@
...
+        Description:
+            This method sends a request to authorize devices in the Plug and Play (PnP) service using the provided
+            device IDs. If the API call is successful and the status code indicates success, it returns the list
+            of device IDs that were authorized. If there is an error during the authorization process, the method
+            logs the error and returns an empty list. In case of an unsuccessful status code, the failure is logged,
+            and an empty list is also returned.
+        """
+
+        payload = {"deviceIdList": device_ids}
+        response = None
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2334]: no need, setting in next line

File: plugins/modules/lan_automation_workflow_manager.py [hunk 7]
```diff
@@ -0,0 +1,2677 @@
...
+        Returns:
+            list: A list of device information if found; an empty list if not found or if an error occurs.
+        Description:
+            This method calls the device onboarding API to fetch information about devices using the provided
+            serial numbers. If the API call is successful and returns valid data, it returns the list of device
+            information. If there is an error or the response is invalid, the method logs the issue and returns
+            an empty list.
+        """
+
+        self.log("!! API Call for dev_list: {}".format(device_serial_numbers))
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2375]: self.log("Initiating API call for device list: {}".format(device_serial_numbers))
respose = None ---> not required...

File: plugins/modules/lan_automation_workflow_manager.py [hunk 8]
```diff
@@ -0,0 +1,2677 @@
...
+        Description:
+            This method starts a new LAN automation session, ensuring that no other session is currently running
+            for the same primary device management IP address. It prepares the necessary parameters for the
+            LAN automation request and sends the initiation call to the appropriate API. If successful, it returns
+            a result dictionary containing the task ID. In case of errors, it logs the errors, sets the operation
+            result to failed, and returns an empty dictionary.
+        """
+        session_to_ip_mapping = self.have.get("session_to_ip_map", {})
+        seed_ip_address = lan_automation_params.get("primaryDeviceManagmentIPAddress")
+
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2414]: self.log("Current session to IP mapping: {}".format(session_to_ip_mapping), "DEBUG")
    self.log("Seed IP address being checked: {}".format(seed_ip_address), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 9]
```diff
@@ -0,0 +1,2677 @@
...
+        if lan_auto_params:
+            self.log("Starting LAN automation with the following parameters: {}".format(lan_auto_params), "DEBUG")
+        else:
+            self.log("No valid parameters found for LAN automation.", "WARNING")
+
+        self.log("Starting LAN automation with parameters: {}".format(lan_auto_params), "DEBUG")
+        lan_auto_params_list = []
+        lan_auto_params_list.append(lan_auto_params)
+
+        try:
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2444]: self.log("Sending request to start LAN Automation with payload: {}".format(lan_auto_params_list), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 10]
```diff
@@ -0,0 +1,2677 @@
...
+            self.log("Response received for starting LAN Automation API call: {}".format(response), "DEBUG")
+
+            if response:
+                self.result.update(dict(response=response["response"]))
+                task_id = response["response"].get("taskId")
+                self.log("Task Id for starting LAN Automation session is {}".format(task_id), "DEBUG")
+                if task_id:
+                    self.result['task_id'] = task_id
+                return self.result
+            else:
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2461]: this else is not required as we return from if response..

       if response:
            self.result.update(dict(response=response["response"]))
            task_id = response["response"].get("taskId")
            self.log("Task ID for starting LAN Automation session is {}".format(task_id), "DEBUG")

            if task_id:
                self.result['task_id'] = task_id

            return self.result

        self.msg = "No response received from the LAN Automation API call."
        self.set_operation_result("failed", False, self.msg, "ERROR")
        self.check_return_status()

File: plugins/modules/lan_automation_workflow_manager.py [hunk 11]
```diff
@@ -0,0 +1,2677 @@
...
+        Returns:
+            self: The current instance of the class, with updated status and messages.
+        Description:
+            This method polls the Cisco Catalyst Center to verify if the specified LAN automation session is still active.
+            If the session is found to be stopped, it logs a success message and updates the operation result.
+            If the session remains active beyond the specified wait time, it logs a timeout message and updates the
+            operation result accordingly.
+        """
+        end_time = time.time() + max_wait_time
+
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2561]: self.log("Starting to check if the LAN automation session is stopped for seed IP '{0}' with session ID '{1}'.".format(seed_ip_address, session_id), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 12]
```diff
@@ -0,0 +1,2677 @@
...
+        Args:
+            config (dict): Configuration details to be verified.
+        Returns:
+            self: The current instance of the class used for interacting with Cisco Catalyst Center.
+        Description:
+            This method checks for an active LAN automation session associated with the primary device management IP address.
+            If found, it logs the session ID; otherwise, it logs a message indicating that no session is active for the
+            specified seed IP.
+        """
+
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2601]: if not self.want.get("lan_automated_device_update"):
        self.log("LAN automated device update is not requested. Exiting verification.", "DEBUG")
        return self

    seed_ip_address = self.want.get("lan_automation").get("primaryDeviceManagmentIPAddress")
    self.log("Verifying active LAN automation session for seed IP: {}".format(seed_ip_address), "DEBUG")

File: plugins/modules/lan_automation_workflow_manager.py [hunk 13]
```diff
@@ -0,0 +1,2677 @@
...
+        return self
+
+
+def main():
+    """
+    main entry point for module execution
+    """
+
+    # Define the specification for the module's arguments
+    element_spec = {"dnac_host": {"required": True, "type": "str"},
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:2631]:
```
   element_spec = {
        "dnac_host": {"required": True, "type": "str"},
         ..
         ..
        'state': {'default': 'merged', 'choices': ['merged', 'deleted']}
    }
```

**Review by madhansansel — COMMENTED** [review]
I added a few comments in the log message. Can you check it and change the appropriate log level?

File: plugins/modules/lan_automation_workflow_manager.py [hunk 14]
```diff
@@ -0,0 +1,2677 @@
...
+            if len(device_details) > 0:
+                hostname = device_details[0].get("hostname")
+                if hostname:
+                    self.log("Hostname for IP {0} is {1}.".format(management_ip_address, hostname), "INFO")
+            else:
+                self.log("No device found with IP address {0}.".format(management_ip_address), "WARNING")
+
+        return hostname
+
+    def get_site_details(self, site_name_hierarchy):
```
[the remark this answers is not in the document]
Comment (maaliaj) [reply plugins/modules/lan_automation_workflow_manager.py:1014]: We have few functions like get_site_id, get_site_type and get_site.
I tried incorporating get_site and check the API response to get the details needed, however I kept getting receiving None as a response so will need to check with the team and see if there is any issues with that function.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 15]
```diff
@@ -0,0 +1,2677 @@
...
+
+        want = self.update_dict_keys_to_camel_case(want)
+
+        self.want = want
+        self.msg = "Successfully collected all parameters from playbook for comparison"
+        self.log("Desired State (want): {0}".format(str(self.want)), "INFO")
+        self.status = "success"
+        return self
+
+    def snake_to_camel(self, snake_str):
```
[the remark this answers is not in the document]
Comment (maaliaj) [reply plugins/modules/lan_automation_workflow_manager.py:1102]: We have camel_to_snake but not snake_to_camel from what I have checked.

[the remark this answers is not in the document]
Comment (madhansansel) [reply plugins/modules/lan_automation_workflow_manager.py:1102]: Then we can move this into dnac.py and inform to our team. Thanks!

File: plugins/modules/lan_automation_workflow_manager.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (madhansansel) [reply plugins/modules/lan_automation_workflow_manager.py:1014]: Sure. You can investigate this later. Please kindly close all other comments and we will push the changes in our branch.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 16]
```diff
@@ -1316,20 +1380,39 @@ def validate_discovery_devices(self, discovery_devices, missing_params):
             - It validates the 'device_site_name_hierarchy' if provided, ensuring the site exists. If the site is
               invalid, it raises an error and halts execution.
         """
+        if discovery_devices is None:
+            self.log("No discovery devices found. Skipping validation for 'discovery_devices'.", "INFO")
+            return
 
         if discovery_devices:
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1387]: As we already checked   in line number 1383: 
`if discovery_devices is None:
`
we don't need this check in 1387: 
`if discovery_devices:`

File: plugins/modules/lan_automation_workflow_manager.py [hunk 17]
```diff
@@ -1345,25 +1428,52 @@ def validate_device_list(self, device_list, missing_params):
             - It checks if the provided IP addresses exist in the system. If not, it raises an error and stops.
             - Missing or invalid parameters are appended to the 'missing_params' list.
         """
+        if not device_list:
+            self.log("No devices to validate in device_list. Skipping validation.", "ERROR")
+            return
+
+        self.log("Starting validation of devices in device list: {}".format(device_list), "DEBUG")
 
         if device_list:
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1437]: Same here. We already checked device_list and returning from 1431. This if check is not required.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 18]
```diff
@@ -1687,64 +1846,134 @@ def verify_diff_merged(self, config):
...
-                        not self.check_link_details(destination_ip_address, destination_interface_name):
-                    self.log("Link between {0}/{1} and {2}/{3} has already been removed.".format(
-                        source_ip_address, source_interface_name, destination_ip_address, destination_interface_name),
-                        "INFO")
-                else:
-                    self.log("Link between {0}/{1} and {2}/{3} still exists and needs to be removed.".format(
-                        source_ip_address, source_interface_name, destination_ip_address, destination_interface_name),
-                        "WARNING")
+    def process_loopback_updates(self, loopback_updates):
+        """Processes loopback IP updates and logs the results.
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1859]: Could you please write the function description on the next line, similar to how you write for other functions? Thank you.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 19]
```diff
@@ -1687,64 +1846,134 @@ def verify_diff_merged(self, config):
...
+        self.log("Processing loopback updates.", "INFO")
+        for loopback in loopback_updates:
+            new_ip = loopback.get("newLoopback0IPAddress")
+            if self.get_ip_details(new_ip, "new loopback IP update"):
+                self.log("Verified loopback IP address {0} was updated on Catalyst Center.".format(new_ip), "INFO")
+            else:
+                self.log("Loopback IP address {0} was not updated on Catalyst Center.".format(new_ip), "WARNING")
+
+    def process_hostname_updates(self, hostname_updates):
+        """Processes hostname updates and logs the results.
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1883]: Could you please write the function description on the next line, similar to how you write for other functions? Thank you.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 20]
```diff
@@ -1687,64 +1846,134 @@ def verify_diff_merged(self, config):
...
+
+            if current_hostname == new_hostname:
+                self.log("Hostname for device IP {0} is already updated to {1}.".format(device_ip, new_hostname),
+                         "INFO")
+            else:
+                self.log("Hostname for device IP {0} was not updated to {1}.".format(device_ip, new_hostname),
+                         "WARNING")
+
+    def process_link_addition(self, link_add):
+        """Processes link addition and logs the results.
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1912]: Could you please write the function description on the next line, similar to how you write for other functions? Thank you.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 21]
```diff
@@ -1687,64 +1846,134 @@ def verify_diff_merged(self, config):
...
+                self.log("Link between {0}/{1} and {2}/{3} was added successfully in Catalyst Center.".format(
+                    source_ip_address, source_interface_name, destination_ip_address, destination_interface_name),
+                    "INFO")
+            else:
+                self.log("Link between {0}/{1} and {2}/{3} was not added in Catalyst Center.".format(
+                    source_ip_address, source_interface_name, destination_ip_address, destination_interface_name),
+                    "WARNING")
+
+    def process_link_deletion(self, link_delete):
+        """Processes link deletion and logs the results.
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1946]: Could you please write the function description on the next line, similar to how you write for other functions? Thank you.

File: plugins/modules/lan_automation_workflow_manager.py [hunk 22]
```diff
@@ -1687,64 +1846,134 @@ def verify_diff_merged(self, config):
...
+        Description:
+            This method checks the details of the link to be added and verifies
+            whether both source and destination links exist in the system. It
+            logs the result of the link addition attempt.
+
+        Returns:
+            None
+        """
+        self.log("Processing link addition.", "INFO")
+        if link_add:
```
Comment (madhansansel) [inline plugins/modules/lan_automation_workflow_manager.py:1929]: You can check link_add, if it is none then you can add a log message and return.. You can also do the same for process_loopback_updates, process_hostname_updates and process_link_deletion..

**Review by madhansansel — APPROVED** [review]

**Review by madhansansel — APPROVED** [review]

**Review by madhansansel — APPROVED** [review]

**Pull request merged**